File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
class Solution {
    
    static int [] array;
    
    static int answer=0;
    
    public void dfs(int depth, int [] numbers,int target){
        if(depth==numbers.length){
            int temp=0;
            for(int i=0;i<numbers.length;i++){
                if(array[i]==1){
                    temp+=numbers[i];
                }
                else{
                    temp-=numbers[i];
                }
            }
            if(temp==target)
                answer+=1;
            return;
        }
        array[depth]=1;
        dfs(depth+1,numbers,target);
        array[depth]=-1;
        dfs(depth+1,numbers,target);
    
    }
    
     
    public int solution(int[] numbers, int target) {
        array=new int[numbers.length];
        dfs(0,numbers,target);
        return answer;
    }
    
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
answer=0
def dfs(numbers,target,calculates,index):
    global answer
    if index==len(numbers):
        temp=0
        for i in range(len(calculates)):
            if calculates[i]=='-':
                temp-=numbers[i]
            else:
                temp+=numbers[i]  
        if temp==target:
            answer+=1
        return
    calculates[index]='-'
    dfs(numbers,target,calculates,index+1)
    calculates[index]='+'
    dfs(numbers,target,calculates,index+1)
    
    

def solution(numbers, target):
    global answer
    cal=[0]*(len(numbers))
    dfs(numbers,target,cal,0)
    return answer
    